### PR TITLE
Rule: ModifierOrder

### LINKED ISSUE #129: Rule: Modifier order
`private lateinit var` vs. `lateinit private var`

Both are valid in Kotlin, but I'd like the ability to enforce one.

### PR DESCRIPTION
**Pull request by Mauin**
Resolves #129 

The modifier order and rule setup was inspired by ktlint: https://github.com/shyiko/ktlint/blob/master/ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ModifierOrderRule.kt#L35

### CODE REVIEW

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+			val findings = rule.lint(file)
+			Assertions.assertThat(findings).isEmpty()
+		}
+	}
+
+	given("a kt parameter with wrongly ordered modifiers") {
+		val code = "lateinit internal private val test: String"
+
+		it("should report the package statement and import statements by default") {
```
Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt]: I think this is a mistake?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+			val findings = rule.lint(code)
+			Assertions.assertThat(findings).hasSize(1)
+		}
+	}
+
+	given("a kt parameter with correctly ordered modifiers") {
+		val code = "private internal lateinit val test: String"
+
+		it("should report the package statement and import statements by default") {
```
Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt]: Here too?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt (hunk 1 above)
Comment (Mauin) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt]: Oh.. Good catch. Copy pasted the MaxLineLengthSpec and forgot to change the description 🙃

**Comment by Mauin** [comment]
Updated the tests and included a Test class for the Modifier order. It felt wrong writing that class with the modifiers all over the place... 😐

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		}
+	}
+
+	given("a kt file with correctly ordered modifiers") {
+		val file = compileForTest(Case.Default.path())
+
+		it("should not report modifiers") {
+			val rule = ModifierOrder()
+
+			val findings = rule.lint(file.text)
```
Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt]: Just give `Case.Default.path` to `lint` or you compile two times the file :)

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+			Assertions.assertThat(findings).isEmpty()
+		}
+	}
+
+	given("a kt file with correctly ordered modifiers") {
+		val file = compileForTest(Case.ModifierOrder.path())
```
Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ModifierOrderSpec.kt]: Here too

**Pull request merged**